ext/opentelemetry-ext-docker-tests/tests/mysql/test_mysql_functional.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2020, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import time
+import unittest
+
+import mysql.connector
+
+from opentelemetry import trace as trace_api
+from opentelemetry.ext.mysql import trace_integration
+from opentelemetry.sdk.trace import Tracer, TracerProvider
+from opentelemetry.sdk.trace.export import SimpleExportSpanProcessor
+from opentelemetry.sdk.trace.export.in_memory_span_exporter import (
+    InMemorySpanExporter,
+)
+
+MYSQL_USER = os.getenv("MYSQL_USER ", "testuser")
+MYSQL_PASSWORD = os.getenv("MYSQL_PASSWORD ", "testpassword")
+MYSQL_HOST = os.getenv("MYSQL_HOST ", "localhost")
+MYSQL_PORT = int(os.getenv("MYSQL_PORT ", "3306"))
+MYSQL_DB_NAME = os.getenv("MYSQL_DB_NAME ", "opentelemetry-tests")
+
+
+class TestFunctionalMysql(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls._tracer_provider = TracerProvider()
+        cls._tracer = Tracer(cls._tracer_provider, None)
+        cls._span_exporter = InMemorySpanExporter()
+        cls._span_processor = SimpleExportSpanProcessor(cls._span_exporter)
+        cls._tracer_provider.add_span_processor(cls._span_processor)
+        trace_integration(cls._tracer)
+        # Try to connect to DB
+        for i in range(5):
+            try:
+                cls._connection = mysql.connector.connect(
+                    user=MYSQL_USER,
+                    password=MYSQL_PASSWORD,
+                    host=MYSQL_HOST,
+                    port=MYSQL_PORT,
+                    database=MYSQL_DB_NAME,
+                )
+                cls._cursor = cls._connection.cursor()
+                break
+            except mysql.connector.Error as err:
+                print(err)
+            time.sleep(5)
```

Comment:
If all 5 attempts fail, then the test execution will continue and then the test cases will fail, probably with non obvious errors like `AttributeError` because `self._cursor` never got defined. This is of course not convenient and probably a bit hard to debug.

This is a situation where there needs to be a separation between test case setup and test case execution. The former are all the operations needed to prepare the test case to be executed, the latter, the execution of the test case itself. The latter is also the one that returns an actual test result, since it is the part where the testing is done. If the setup fails, the execution should not be performed because it makes no sense to run a test case when its setup is not ready.

As far as I know, Pyhton `unittest` package does not support this, for example:

```python
from unittest import TestCase


class TestCase(TestCase):

    def setUp(self):
        1 / 0

    def test_0(self):
        assert True
```
```
============================= test session starts ==============================
platform linux -- Python 3.6.9, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/ocelotl/sandbox/tast
collected 1 item                                                               

test_unittest.py F                                                       [100%]

=================================== FAILURES ===================================
_______________________________ TestCase.test_0 ________________________________

self = <test_unittest.TestCase testMethod=test_0>

    def setUp(self):
>       1 / 0
E       ZeroDivisionError: division by zero

test_unittest.py:7: ZeroDivisionError
=========================== short test summary info ============================
FAILED test_unittest.py::TestCase::test_0 - ZeroDivisionError: division by zero
============================== 1 failed in 0.08s ===============================
```

Here, a *failure* is reported, nevertheless this is not right because a failure means there is something wrong with the code under test, and this code never got tested because the test case never got executed.

This is why Pytest supports the concept of *error* when reporting testing results, for example:
```python
from pytest import fixture


@fixture
def the_fixture():
    1 / 0


def test_0(the_fixture):
    assert True
```
```
============================= test session starts ==============================
platform linux -- Python 3.6.9, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/ocelotl/sandbox/tast
collected 1 item                                                               

test_pytest.py E                                                         [100%]

==================================== ERRORS ====================================
___________________________ ERROR at setup of test_0 ___________________________

    @fixture
    def the_fixture():
>       1 / 0
E       ZeroDivisionError: division by zero

test_pytest.py:6: ZeroDivisionError
=========================== short test summary info ============================
ERROR test_pytest.py::test_0 - ZeroDivisionError: division by zero
=============================== 1 error in 0.08s ===============================
```

This is much more accurate, because a failure is not reported, but an error is, meaning that something went wrong when preparing the test to be executed. This concept of error is very useful, since it also applies to the teardown of test cases:
```python
from pytest import fixture


@fixture
def the_fixture():
    # A pytest fixture uses yield to separate the setup from the teardown
    pass  # setup
    yield
    1 / 0  # teardown


def test_0(the_fixture):
    assert True
```
```
============================= test session starts ==============================
platform linux -- Python 3.6.9, pytest-5.4.1, py-1.8.1, pluggy-0.13.1
rootdir: /home/ocelotl/sandbox/tast
collected 1 item

test_pytest.py .E                                                        [100%]

==================================== ERRORS ====================================
_________________________ ERROR at teardown of test_0 __________________________

    @fixture
    def the_fixture():
        # A pytest fixture uses yield to separate the setup from the teardown
        pass  # setup
        yield
>       1 / 0  # teardown
E       ZeroDivisionError: division by zero

test_pytest.py:9: ZeroDivisionError
=========================== short test summary info ============================
ERROR test_pytest.py::test_0 - ZeroDivisionError: division by zero
========================== 1 passed, 1 error in 0.02s ==========================
```

This is very accurate test reporting, the test case was executed, the code under test tested and it passed. This is reported accurately with `1 passed`. Nevertheless, something went wrong when running the test teardown, and this is also accurately reported with `1 error`.

I realize that we have been using the `unittest.TestCase` style of test case writing so far, but I think we have hit a limitation of this style now with a test case that requires this kind of setup where an external component may fail, rendering the rest of the testing useless. It would be ideal, of course, to be able to use fixtures with our current `unittest.TestCase` test cases in order to maintain consistency with the rest of the existing test cases, but sadly, this [does not seem to be possible](https://docs.pytest.org/en/latest/unittest.html#pytest-features-in-unittest-testcase-subclasses):

> The following pytest features do not work, and probably never will due to different design philosophies:
> Fixtures (except for autouse fixtures, see below);

The autouse fixtures are fixtures that are automatically applied to a complete set of test cases, this may be a solution for this specific situation, but probably this may not work always because of them being applied to all test cases in the set.

Please reach out to the rest of the team if you are unsure about my suggestions here, I understand that I am asking to use Pytest fixtures and non `unittest.TestCase` test cases which can be a bit controversial.

Thanks and sorry for the long post!

Reply:
Added pytest fixture for DB connection keeping unittest.TestCase structure, if we don't need to refactor the entire tests we currently have I believe is ok to use pytest features.

Reply:
Thanks for the explanation @ocelotl, very helpful.

Some reasons not to use fixtures:
- they make it impossible to run tests without pytest
- they're magical -- non-autouse fixtures add a phantom arg to test methods and make it hard to reason about the behavior of tests
- autouse fixtures tend to be used for heavyweight background work -- like starting a docker instance -- that can make tests slow and unreliable

The bad may not outweigh the good here, and this does look for a good case for a fixture. I know it's possible to make a mess out of tests even without fixtures, it's just that fixtures makes that especially easy.

Reply:
Maybe we can talk more during the SIG about this? I would say a mixed unittest / pytest fixture pattern is the most confusing, as it's the two patterns combined and requires an understanding of how both work.

The ERROR / FAILURE distinction, albeit semantically clearer, don't seem that important IMO. Either way I need to go look at the code that broke, so I'm not sure what that changes from my actions.